### PR TITLE
fix(localize): allow ICU expansion to start with any character except…

### PR DESCRIPTION
**Pull request by sonukapoor**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #31586

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/compiler/src/ml_parser/lexer.ts [hunk 1]
```diff
@@ -747,7 +747,7 @@ function isNamedEntityEnd(code: number): boolean {
 }
 
 function isExpansionCaseStart(peek: number): boolean {
-  return peek === chars.$EQ || chars.isAsciiLetter(peek) || chars.isDigit(peek);
+  return peek === chars.$EQ || peek !== chars.$RBRACE;
```
Comment (petebacondarwin) [inline packages/compiler/src/ml_parser/lexer.ts]:
```suggestion
  return peek !== chars.$RBRACE;
```

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
No need for the additional check now.

In the commit message, please change references to `expansion` to `expansion case` since this is what the commit is fixing.

File: packages/compiler/src/ml_parser/lexer.ts (hunk 1 above)
Comment (sonukapoor) [reply packages/compiler/src/ml_parser/lexer.ts]: Yes, of course...

**Review by petebacondarwin — APPROVED** [review]
Great! Thanks

**Comment by AndrewKushnir** [comment]
Hi @sonukapoor, sorry I didn't have a chance to look at the changes in this PR today. Even though the diff is relatively small, I'd like to understand the code around it and see if there are any corner cases and why the function was not implemented like this before. I've started [g3 presubmit](http://test/OCL:301691009:BASE:301691112:1584573408494:76d9f45a) (internal only link) for this change as well to see if there are any cases in g3 that might be affected. Thank you.

**Comment by sonukapoor** [comment]
Hi @AndrewKushnir, no problem at all:

I have discussed this offline with @petebacondarwin to make sure that we are safe to change this.

The template in question is:

```
<span i18n>{
  gOptionLabel, select,
  National transport {National transport}
  % by which guarantee amount is to be reduced {% by which guarantee amount is to be reduced}
  }
</span>
```

Once the expansion has started, we continue until we find the closing character `}`. In between that we only allow for alphanumeric characters. However, in the above example, we can see that `%` is being used instead, other characters could be used as well.  

To allow the usage of any character, except for the closing right bracket `}`, we have changed the condition. This resolves the issue and the current existing tests are still passing as well. Once, we hit the end character `}`, we close the expansion and can move on. This logic can be seen in this function (added comment blocks below only to make it easier to follow);

```
private _tokenizeExpansionForm(): boolean {
  // Expansion starts here
  if (this.isExpansionFormStart()) {
    this._consumeExpansionFormStart();
    return true;
  }

  // find the next char, and allow anything except `}`
  if (isExpansionCaseStart(this._cursor.peek()) && this._isInExpansionForm()) {
    this._consumeExpansionCaseStart();
    return true;
  }

  // Close the expansion as the current char is `}`
  if (this._cursor.peek() === chars.$RBRACE) {
    if (this._isInExpansionCase()) {
      this._consumeExpansionCaseEnd();
      return true;
    }

    if (this._isInExpansionForm()) {
      this._consumeExpansionFormEnd();
      return true;
    }
  }
  return false;
}
```

[Here](https://github.com/angular/angular/commit/a3a54299afd41cc2049953c52ff869ef9c36fc37)is a previous commit from Olivier that added support for numerical digits in expansion cases. 

@petebacondarwin please add anything else  I may have missed.

**Comment by petebacondarwin** [comment]
@AndrewKushnir - yes @sonukapoor and I spent a lot of time discussing this offline and working through the possibilities. The [ICU spec](http://userguide.icu-project.org/formatparse/messages) is pretty vague:

> Use a "select" argument to select sub-messages via a fixed set of keywords.

It does not specify what can be a "keyword" but from looking at the surrounding syntax it appears that it can indeed be any string that does not contain a `}` character.

**Review by AndrewKushnir — APPROVED** [review]
Thanks for additional information @sonukapoor and @petebacondarwin, the change LGTM 👍 

The g3 presubmit is successful, I will start a global run tonight as well to test all affected targets in g3 just to make sure everything is ok.

Thank you.

**Comment by sonukapoor** [comment]
Thanks 🙏

**Comment by AndrewKushnir** [comment]
[Link to Global TAP run](https://test.corp.google.com/ui#id=OCL:301691009:BASE:301943198:1584670359389:bfb2773f) (internal only).

**Comment by AndrewKushnir** [comment]
@sonukapoor, @petebacondarwin global presubmit looks good 👍

I have a couple more questions/notes:

- can we extend commit message to include info that Pete mentioned in https://github.com/angular/angular/pull/36123#issuecomment-601074010, that'd help to provide more context for future references

- previously whitespace symbols were not treated as a possible ICU case start. With this change whitespace symbols can technically be treated as ICU case start. Can we check if tis change doesn't affect i18n extraction mechanism and that i18n message IDs remain the same for ICUs after this change (otherwise it would trigger re-translation for all messages that contain ICUs)? Can we add a test (or find existing ones) to verify that?

Sorry for additional questions, just want to make sure we are not missing anything.

Thank you.

**Comment by petebacondarwin** [comment]
Hi @AndrewKushnir 
I believe that we covered with regards to whitespace by this line: https://github.com/angular/angular/blob/f78876695dd46f28b99246536f3503d5e452bdb5/packages/compiler/src/ml_parser/lexer.ts#L624

When we consume an expansion case start we also trim whitespace off the front and back.
The only scenario I can think of would be to incorrectly identify a block of "only whitespace" as an expansion case. But this is actually covered by this line: https://github.com/angular/angular/blob/f78876695dd46f28b99246536f3503d5e452bdb5/packages/compiler/src/ml_parser/lexer.ts#L619 and this line https://github.com/angular/angular/blob/f78876695dd46f28b99246536f3503d5e452bdb5/packages/compiler/src/ml_parser/lexer.ts#L640

where consuming an "expansion form start" and an "expansion case" always end by skipping over whitespace.

**Comment by sonukapoor** [comment]
@AndrewKushnir I have updated the commit message.

**Comment by AndrewKushnir** [comment]
Thanks for additional information and updates @sonukapoor and @petebacondarwin, I'm marking this PR as ready for merge. Thank you.
